README.md
# cupertino_rounded_corners

A widget and border to make cupertino rounded corners also refrered to as squircles using a bezier path and having the two points in the corners.

## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).



## Import it

```
import 'package:cupertino_rounded_corners/cupertino_rounded_corners.dart';
```

## To make your own card
```
new Container(
  decoration: new BoxDecoration(
    color: Colors.transparent,
    borderRadius: new BorderRadius.all(
      new Radius.circular(4.0),
    ),
  ),
  margin: margin,
  child: new Material(
    color: color,
    shape: new SquircleBorder(
      radius: BorderRadius.all(
        new Radius.circular(40.0),
      ),
    ),
    elevation: elevation,
    child: new Padding(padding: padding, child: child),
  ),
);
```

## Use the card
```
CupertinoCard(
  child: const Text("HELLO WORLD"),
  elevation: 2.0,
  margin: const EdgeInsets.all(4.0),
  padding: const EdgeInsets.all(0.0),
  color: Colors.white,
  radius: BorderRadius.all(
    new Radius.circular(40.0),
  ),
);
```

To achieve the bottom last button in the second example.
```
CupertinoCard(
  splashColor: Colors.green,
  child: Container(
    constraints: BoxConstraints.expand(
      height: 100.0,
    ),
    child: Padding(
      padding: EdgeInsets.all(8.0),
      child: const Text(
        'This is a button',
        style: TextStyle(color: Colors.white),
      ),
    ),
  ),
  decoration: BoxDecoration(
    gradient: LinearGradient(
      begin: Alignment.topLeft,
      end: Alignment.bottomRight,
      stops: [0.0, 0.2, 0.8, 1.0],
      colors: [
        Color.fromRGBO(240, 40, 180, 1.0),
        Color.fromRGBO(220, 40, 200, 1.0),
        Color.fromRGBO(200, 40, 220, 1.0),
        Color.fromRGBO(180, 40, 240, 1.0),
      ],
    ),
  ),
  elevation: 2.0,
  margin: const EdgeInsets.all(4.0),
  padding: const EdgeInsets.all(0.0),
  radius: BorderRadius.all(
    new Radius.circular(40.0),
  ),
  onPressed: () {
    print("Clicked");
  },
);
```

## Example Image Mobile
<!--![Alt text](screenshots/screenshot1.png?raw=true "Title")-->
![alt text](https://github.com/monksc/cupertino_rounded_corners/blob/master/screenshots/screenshot1.png?raw=true)

## Example Image Linux
<!--![Alt text](screenshots/screenshot1.png?raw=true "Title")-->
![alt text](https://github.com/monksc/cupertino_rounded_corners/blob/master/screenshots/screenshot2.png?raw=true)



CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.

## [1.0.2] - TODO: Add release date.

Made bug fixes

## [1.0.3] - TODO: Add release date.

Updated README to show correct usage and added support for
a radius in the card.


## [1.1.0] - TODO: Add release date.

Added in examples and now all radius use a BorderRadius object.


## [1.1.3] - TODO: Add release date.

Added in screenshots to the README and added them to the README.

## [2.0.0] - TODO: Add release date.

Updated code to support Flutter 2.0 null safety.

## [2.0.1] - Null Safety

Fixed a few descriptions and some other small things.

## [2.1.0] - Gradient Background

CupertinoCard now supports gradient backgrounds.
CupertinoCard can be tapped and contains the widget InkWell.
CupertinoCard now has a clipped path so the child is clipped.



example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

// import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

// import 'package:example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    expect(0, 0);
    // // Build our app and trigger a frame.
    // await tester.pumpWidget(MyApp());

    // // Verify that our counter starts at 0.
    // expect(find.text('0'), findsOneWidget);
    // expect(find.text('1'), findsNothing);

    // // Tap the '+' icon and trigger a frame.
    // await tester.tap(find.byIcon(Icons.add));
    // await tester.pump();

    // // Verify that our counter has incremented.
    // expect(find.text('0'), findsNothing);
    // expect(find.text('1'), findsOneWidget);
  });
}


test/cupertino_rounded_corners_test.dart
import 'package:test/test.dart';
// import 'package:cupertino_rounded_corners/cupertino_rounded_corners.dart';

void main() {
  test('adds one to input values', () {
    expect(0, 0);
  });
}


lib/cupertino_rounded_corners.dart
import 'package:flutter/material.dart';

Path SquirclePath(Rect rect, BorderRadius? radius) {
  final c = rect.center;
  double startX = rect.left;
  double endX = rect.right;
  double startY = rect.top;
  double endY = rect.bottom;

  double midX = c.dx;
  double midY = c.dy;

  if (radius == null) {
    return new Path()
      ..moveTo(startX, midY)
      ..cubicTo(startX, startY, startX, startY, midX, startY)
      ..cubicTo(endX, startY, endX, startY, endX, midY)
      ..cubicTo(endX, endY, endX, endY, midX, endY)
      ..cubicTo(startX, endY, startX, endY, startX, midY)
      ..close();
  }

  return new Path()

    // Start position
    ..moveTo(startX, startY + radius.topLeft.y)

    // top left corner
    ..cubicTo(
      startX,
      startY,
      startX,
      startY,
      startX + radius.topLeft.x,
      startY,
    )

    // top line
    ..lineTo(endX - radius.topRight.x, startY)

    // top right corner
    ..cubicTo(
      endX,
      startY,
      endX,
      startY,
      endX,
      startY + radius.topRight.y,
    )

    // right line
    ..lineTo(endX, endY - radius.bottomRight.y)

    // bottom right corner
    ..cubicTo(
      endX,
      endY,
      endX,
      endY,
      endX - radius.bottomRight.x,
      endY,
    )

    // bottom line
    ..lineTo(startX + radius.bottomLeft.x, endY)

    // bottom left corner
    ..cubicTo(
      startX,
      endY,
      startX,
      endY,
      startX,
      endY - radius.bottomLeft.y,
    )

    // left line
    //..moveTo(startX, startY + radius)
    ..close();
}

class SquircleBorder extends ShapeBorder {
  final BorderSide side;
  final BorderRadius? radius;

  const SquircleBorder({
    this.side: BorderSide.none,
    this.radius,
  });

  @override
  EdgeInsetsGeometry get dimensions => EdgeInsets.all(side.width);

  @override
  ShapeBorder scale(double t) {
    return new SquircleBorder(
      side: side.scale(t),
      radius: radius,
    );
  }

  @override
  Path getInnerPath(Rect rect, {TextDirection? textDirection}) {
    return SquirclePath(rect.deflate(side.width), radius);
  }

  @override
  Path getOuterPath(Rect rect, {TextDirection? textDirection}) {
    return SquirclePath(rect, radius);
  }

  @override
  void paint(Canvas canvas, Rect rect, {TextDirection? textDirection}) {
    switch (side.style) {
      case BorderStyle.none:
        break;
      case BorderStyle.solid:
        var path = getOuterPath(rect.deflate(side.width / 2.0),
            textDirection: textDirection);
        canvas.drawPath(path, side.toPaint());
    }
  }
}

class CupertinoCard extends StatelessWidget {
  final EdgeInsets margin;
  final EdgeInsets padding;
  final Widget? child;
  final double elevation;
  final Color color;
  final Color? splashColor;
  final BorderRadius radius;
  final Decoration? decoration;
  final VoidCallback? onPressed;

  CupertinoCard({
    this.child,
    this.elevation = 2.0,
    this.margin: const EdgeInsets.all(4.0),
    this.padding: const EdgeInsets.all(0.0),
    this.color: Colors.white,
    this.splashColor,
    this.decoration,
    this.radius: const BorderRadius.all(
      const Radius.circular(
        40.0,
      ),
    ),
    this.onPressed,
  });

  @override
  Widget build(BuildContext context) {
    var shapeborder = SquircleBorder(
      radius: radius,
    );

    return new Padding(
      padding: margin,
      child: new Material(
        elevation: elevation,
        shape: shapeborder,
        child: ClipPath.shape(
          shape: shapeborder,
          child: Material(
            color: color,
            child: Ink(
              decoration: decoration,
              child: InkWell(
                customBorder: shapeborder,
                onTap: onPressed,
                splashColor: splashColor,
                child: new Padding(
                  padding: padding,
                  child: child,
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:cupertino_rounded_corners/cupertino_rounded_corners.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Welcome to Flutter',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Cupertino Rounded Corners'),
        ),
        body: ListView(
          shrinkWrap: true,
          padding: const EdgeInsets.all(20.0),
          children: <Widget>[
            new Container(
              decoration: new BoxDecoration(
                color: Colors.transparent,
                borderRadius: new BorderRadius.all(
                  new Radius.circular(4.0),
                ),
              ),
              margin: new EdgeInsets.all(8.0),
              child: new Material(
                color: Colors.white,
                shape: new SquircleBorder(
                  radius: BorderRadius.all(
                    Radius.circular(
                      20.0,
                    ),
                  ),
                ),
                elevation: 8.0,
                child: new Padding(
                  padding: new EdgeInsets.all(8.0),
                  child: Text('HELLO WORLD'),
                ),
              ),
            ),
            new Container(
              decoration: new BoxDecoration(
                color: Colors.transparent,
                borderRadius: new BorderRadius.all(
                  new Radius.circular(4.0),
                ),
              ),
              margin: new EdgeInsets.all(8.0),
              child: new Material(
                color: Colors.cyan,
                shape: new SquircleBorder(
                  radius: BorderRadius.all(
                    Radius.elliptical(
                      40.0,
                      40.0,
                    ),
                  ),
                ),
                elevation: 8.0,
                child: new Padding(
                  padding: new EdgeInsets.all(30.0),
                  child: Text(
                    'Bigger and Colorful',
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ),
            ),
            new Container(
              decoration: new BoxDecoration(
                color: Colors.transparent,
                borderRadius: new BorderRadius.all(
                  new Radius.circular(4.0),
                ),
              ),
              margin: new EdgeInsets.all(8.0),
              child: new Material(
                color: Colors.red,
                shape: new SquircleBorder(
                  radius: BorderRadius.all(
                    Radius.elliptical(
                      250.0,
                      40.0,
                    ),
                  ),
                ),
                elevation: 8.0,
                child: new Padding(
                  padding: new EdgeInsets.all(50.0),
                  child: Text(
                    'More curved horizontally',
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ),
            ),
            new Container(
              decoration: new BoxDecoration(
                color: Colors.transparent,
                borderRadius: new BorderRadius.all(
                  new Radius.circular(4.0),
                ),
              ),
              margin: new EdgeInsets.all(8.0),
              child: new Material(
                color: Colors.red,
                shape: new SquircleBorder(
                  radius: BorderRadius.vertical(
                    top: Radius.elliptical(
                      250.0,
                      100.0,
                    ),
                  ),
                ),
                elevation: 8.0,
                child: new Padding(
                  padding: new EdgeInsets.all(50.0),
                  child: Text(
                    'Top is curved only',
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ),
            ),
            new Container(
              decoration: new BoxDecoration(
                color: Colors.transparent,
                borderRadius: new BorderRadius.all(
                  new Radius.circular(4.0),
                ),
              ),
              margin: new EdgeInsets.all(8.0),
              child: new Material(
                color: Colors.green,
                shape: new SquircleBorder(
                  radius: BorderRadius.vertical(
                    bottom: Radius.elliptical(
                      250.0,
                      100.0,
                    ),
                  ),
                ),
                elevation: 8.0,
                child: new Padding(
                  padding: new EdgeInsets.all(50.0),
                  child: Text(
                    'Bottom is curved only',
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ),
            ),
            Align(
              alignment: Alignment.center,
              child: Container(
                width: 250,
                height: 220,
                child: new Container(
                  child: new Material(
                    color: Colors.blue,
                    shape: new SquircleBorder(
                      radius: BorderRadius.all(
                        new Radius.circular(170.0),
                      ),
                    ),
                    elevation: 8.0,
                    child: new Padding(
                      padding: new EdgeInsets.all(50.0),
                      child: Text(
                        'Curved all the way',
                        style: TextStyle(color: Colors.white),
                      ),
                    ),
                  ),
                ),
              ),
            ),
            CupertinoCard(
              splashColor: Colors.green,
              child: Container(
                constraints: BoxConstraints.expand(
                  height: 100.0,
                ),
                child: Padding(
                  padding: EdgeInsets.all(8.0),
                  child: const Text(
                    'This is a button',
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ),
              decoration: BoxDecoration(
                // Box decoration takes a gradient
                gradient: LinearGradient(
                  // Where the linear gradient begins and ends
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  // Add one stop for each color. Stops should increase from 0 to 1
                  stops: [0.0, 0.2, 0.8, 1.0],
                  colors: [
                    Color.fromRGBO(240, 40, 180, 1.0),
                    Color.fromRGBO(220, 40, 200, 1.0),
                    Color.fromRGBO(200, 40, 220, 1.0),
                    Color.fromRGBO(180, 40, 240, 1.0),
                  ],
                ),
              ),
              elevation: 2.0,
              margin: const EdgeInsets.all(4.0),
              padding: const EdgeInsets.all(0.0),
              radius: BorderRadius.only(
                topLeft: new Radius.circular(40.0),
                topRight: new Radius.circular(40.0),
                bottomLeft: new Radius.circular(40.0),
                bottomRight: new Radius.circular(40.0),
              ),
              onPressed: () {
                print("Clicked");
              },
            ),
          ],
        ),
      ),
    );
  }
}
